docs/api-docs/headless/end-to-end-guide/graphql.mdx

```diff
@@ -0,0 +1,756 @@
+</Callout>
+
+## Prerequisites
+
+To complete the guide, you will need the following:
+* [BigCommerce sandbox store](https://start.bigcommerce.com/developer-sandbox/) 
+* Familiarity with creating a channel and a site
+
+## Accessing Storefront GraphQL APIs
+You can make BigCommerce GraphQL requests in the context of the server to server API. The server-side API clients have the option of using either a Storefront API token or a customer impersonation token to access Storefront GraphQL APIs. This tutorial describes how to create and use a customer impersonation token to authenticate and access the Storefront GraphQL APIs. 
+
+<Callout type="info">
+While the server-side API clients can use either a Storefront API token or a customer impersonation token, client-side web browsers are not allowed with customer impersonation tokens.
+</Callout>
+
+### Create a customer impersonation token
+
+A customer impersonation token permits you to perform actions as any shopper. To generate a [customer impersonation token](/docs/rest-authentication/tokens#customer-impersonation-tokens), send a POST request to `/v3/storefront/api-token-customer-impersonation`.
+
+<Tabs items={['Request', 'Response']}>
+  <Tab>
+ ```http filename="Example request: Create a customer impersonation token"
+POST https://api.bigcommerce.com/stores/{STORE_HASH}/v3/storefront/api-token-customer-impersonation
+X-Auth-Token: {ACCESS_TOKEN}
+Accept: application/json
+Content-Type: application/json
+
+{
+  "channel_id": 1, 
+  "expires_at": 1602288000 
+}
+```
+</Tab>
+  <Tab>
+
+```json filename="Example response: Create a customer impersonation token"
+  {
+    "data":
+      {
+        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
+      }
+      "meta": {}
+  }
+```
+</Tab>
+</Tabs>
+
+In the GraphQL Storefront Playground, enter the customer impersonation token in the **Headers** tab as follows:
+
+```
+{
+  "Authorization": "Bearer {customer impersonation token}"
+}
+```
+
+## Querying product data
+
+The GraphQL Storefront API lets you retrieve product data from a store. The following example demonstrates how to query a simple product using the GraphQL Storefront API. For more information and examples, see the [Guide to Working with Products](/docs/storefront/graphql/examples/products).
+
+<Callout type="info">
+#### Query best practices
+* Query less than 50 resources at a time.
+* Limit query complexity where possible.
+* Limit pagination where possible.
+* Employ security practices and limit scope where possible.
+* Employ caching where possible.
+</Callout>
+
+```graphql filename="Get a product" showLineNumbers copy
+# This query retrieves one product.
+query SingleProduct {
```

Comment:
Not sure if this will be the case in other example queries in this article, but I'll just note it once here:  

It's confusing that this example query seems to be tailored specifically for a JavaScript context where the `${params.id}` syntax would interpolate a value into the string, and relying on some value called `params.id` that is pre-existing. I would think we'd want to keep the syntax to just GraphQL, agnostic to any specific context:

```
# This query retrieves one product.
query SingleProduct($entityId: Int) {
  site {
    product (entityId: $entityId) {
      id
      entityId
      name
      sku
      description
      prices {
        price {
          value
          currencyCode
        }
      }
    }
  }
}
```